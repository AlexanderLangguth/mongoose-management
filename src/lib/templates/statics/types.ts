/**
 * ######################################################################
 * #                                                                    #
 * #                       Do not change the file!                      #
 * #                                                                    #
 * # All changes are deleted the next time the collection data is read. #
 * #                                                                    #
 * ######################################################################
 */

import { IndexOptions } from 'mongodb';
import { SchemaTypeOpts } from 'mongoose';

/**
 * Helpers
 */
type Keys<T> = { [K in keyof T]: T[K] extends never ? never : K }[keyof T];
type Diff<T, U> = T extends U ? never : T;

/**
 * Extracts the type from the function
 */
type virtualFunctionTypeGet<T extends { [k: string]: any }> = T['get'] extends () => void
  ? ReturnType<T['get']>
  : never;
type virtualFunctionTypeSet<T extends { [k: string]: any }> = T['set'] extends (value: infer A) => void ? A : never;

/**
 * List of all types (with the type "never")
 */
type virtualFunctionTypeGetList<T> = { [K in keyof T]: virtualFunctionTypeGet<T[K]> };
type virtualFunctionTypeSetList<T> = { [K in keyof T]: virtualFunctionTypeSet<T[K]> };

/**
 * List of all types (without the type "never")
 */
type virtualTypeGetList<T> = Pick<virtualFunctionTypeGetList<T>, Keys<virtualFunctionTypeGetList<T>>>;
type virtualTypeSetList<T> = Pick<virtualFunctionTypeSetList<T>, Keys<virtualFunctionTypeSetList<T>>>;

/**
 * List of Keys where the object have only "get"
 */
type virtualKeysGetOnly<T> = Diff<keyof virtualTypeGetList<T>, keyof virtualTypeSetList<T>>;

/**
 * List of Keys where without "get only" objects
 */
type virtualKeysBoth<T> = Diff<keyof T, virtualKeysGetOnly<T>>;

export type virtualType<T> = {
  readonly [K in keyof virtualTypeGetList<T> & virtualKeysGetOnly<T>]: virtualTypeGetList<T>[K];
} &
  { [K in keyof virtualTypeGetList<T> & virtualKeysBoth<T>]: virtualTypeGetList<T>[K] } &
  { [K in keyof virtualTypeSetList<T>]: virtualTypeSetList<T>[K] };

/**
 *
 */
export type extendTimestampType = {
  readonly createdAt: Date;
  readonly updatedAt: Date;
};

/**
 *
 */
export type extendIdType = {
  readonly _id: any;
  readonly id: any;
};

/**
 *
 */
export type indexType = { fields: { [k: string]: number }; options?: IndexOptions };

/**
 *
 */
export type schemaExtendType<T extends Record<string, any>> = {
  [K in keyof T]?: T[K] extends Array<schemaExtendType<T>> ? [schemaExtendType<T[K][0]>] : SchemaTypeOpts<any>;
};
