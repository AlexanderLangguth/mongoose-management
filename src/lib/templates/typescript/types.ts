/**
 * ######################################################################
 * #                                                                    #
 * #                       Do not change the file!                      #
 * #                                                                    #
 * # All changes are deleted the next time the collection data is read. #
 * #                                                                    #
 * ######################################################################
 */

namespace Virtual {
  /**
   * Helpers
   */
  type Keys<T> = { [K in keyof T]: T[K] extends never ? never : K }[keyof T];
  type Diff<T, U> = T extends U ? never : T;

  /**
   * Extracts the type from the function
   */
  type functionTypeGet<T extends { [k: string]: any }> = T['get'] extends Function ? ReturnType<T['get']> : never;
  type functionTypeSet<T extends { [k: string]: any }> = T['set'] extends (value: infer A) => void ? A : never;

  /**
   * List of all types (with the type "never")
   */
  type functionTypeGetList<T> = { [K in keyof T]: functionTypeGet<T[K]> };
  type functionTypeSetList<T> = { [K in keyof T]: functionTypeSet<T[K]> };

  /**
   * List of all types (without the type "never")
   */
  type typeGetList<T> = Pick<functionTypeGetList<T>, Keys<functionTypeGetList<T>>>;
  type typeSetList<T> = Pick<functionTypeSetList<T>, Keys<functionTypeSetList<T>>>;

  /**
   * List of Keys where the object have only "get"
   */
  type keysGetOnly<T> = Diff<keyof typeGetList<T>, keyof typeSetList<T>>;

  /**
   * List of Keys where without "get only" objects
   */
  type keysBoth<T> = Diff<keyof T, keysGetOnly<T>>;

  export type types<T> = { readonly [K in keyof typeGetList<T> & keysGetOnly<T>]: typeGetList<T>[K] } &
    { [K in keyof typeGetList<T> & keysBoth<T>]: typeGetList<T>[K] } &
    { [K in keyof typeSetList<T>]: typeSetList<T>[K] };
}

export type virtualType<T> = Virtual.types<T>;

export type extendTimestampType = {
  readonly createdAt: Date;
  readonly updatedAt: Date;
};

export type extendIdType = {
  readonly _id: any;
  readonly id: any;
};
