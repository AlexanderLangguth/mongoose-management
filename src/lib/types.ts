import { Separator } from 'inquirer';

import Prompts from './prompts';
import Storage from './storage';

import Collection from './cli/sequences/collection';
import Collections from './cli/sequences/collections';
import Column from './cli/sequences/column';
import Index from './cli/sequences/index';

import CollectionAction from './cli/actions/collection';
import ColumnAction from './cli/actions/column';
import GroupAction from './cli/actions/group';
import IndexAction from './cli/actions/index';

import Create from './template/create';

import * as mongoTypes from './mongo';

/**
 *
 */
export type objType<K extends string | number | symbol = string, T = any> = Record<K, T>;

/**
 *
 */
export type cliOptionsType = {
  prompts: Prompts;
  storage: Storage;

  // sequences
  Collections: typeof Collections;
  Collection: typeof Collection;
  Column: typeof Column;
  Index: typeof Index;

  // actions
  actionGroup: GroupAction;
  actionCollection: CollectionAction;
  actionColumn: ColumnAction;
  actionIndex: IndexAction;

  createTemplate: Create;

  data: dataType;
};

/**
 *
 */
export type schemaType = keyof typeof mongoTypes.schemaTypes;

/**
 *
 */
export type schemaNormalType = keyof typeof mongoTypes.schemaTypesNormal;

/**
 *
 */
export type schemaSpecialType = keyof typeof mongoTypes.schemaTypesSpecial;

/**
 *
 */
export type schemaIndexType = keyof typeof mongoTypes.schemaIndexTypes;

/**
 *
 */
export type dataType = {
  groups: dataGroupType[];
};

/**
 *
 */
export type dataGroupType = {
  path: string;
  collections: dataCollectionType[];
};

/**
 *
 */
export type dataCollectionType = {
  name: string;
  columns: dataColumnType[];
  indexes: dataIndexType[];
};

/**
 *
 */
export type dataColumnType = {
  name: string;
  type: schemaType;

  required: boolean;
  default?: string;

  // string
  lowercase?: boolean;
  uppercase?: boolean;
  trim?: boolean;
  match?: string;
  enum?: string;
  minlength?: number;
  maxlength?: number;

  // number
  min?: number;
  max?: number;

  // internal values
  subType?: dataColumnArrayType;
  subColumns?: dataColumnType[];
  showType?: string;
  readonly?: boolean;
};

/**
 *
 */
export type dataColumnArrayType = {
  type: schemaNormalType;
  subType?: dataColumnArrayType;
};

/**
 *
 */
export type dataIndexType = {
  name: string;
  columns: { [k: string]: dataIndexColumnValueType };
  properties: {
    unique?: boolean;
    sparse?: boolean;
    /*
    expires?: number;
    artialFilterExpression?: any;
    collation?: {
      locale: string;
      strength: number;
    };
    */
  };
  // internal values
  readonly?: boolean;
  mode?: schemaIndexType;
  type?: dataIndexColumnValueType;
};

/**
 *
 */
export type dataColumnGroupType = {
  parent?: dataColumnGroupType;
  column: dataColumnType;
};

/**
 *
 */
export type dataIndexColumnValueType = 1 | -1 | 'text' | 'hashed' | '2dsphere';

/**
 *
 */
export type choicesType<R> = choiceType<R> | InstanceType<typeof Separator>;

/**
 *
 */
export type choiceType<R> = choiceListType<choiceValueType<R>>;

/**
 *
 */
export type choiceListType<T> = {
  name: string;
  value: T;
  short: string;
};

/**
 *
 */
export type choiceValueType<R> = {
  type?: 'create' | 'back' | 'exit' | 'save' | 'edit' | 'remove' | 'write';
  value?: R;
};

/**
 *
 */
export type choiceValueSubType =
  | {
      type: 'column';
      data?: dataColumnType;
    }
  | {
      type: 'index';
      data?: dataIndexType;
    };

/**
 *
 */
export type templateTypesType = 'index' | 'documents' | 'interfaces' | 'models' | 'repositories';

/**
 *
 */
export type templateCollectionNamesType = {
  collectionNameLower: string;
  collectionNameUpper: string;
  collectionNameRaw: string;
};

/**
 *
 */
export type templateCollectionType = {
  interfaceName: string;

  schemaDefinitions: string;
  SchemaIndexes: string;
  schemaTypes: string;
} & templateCollectionNamesType;
