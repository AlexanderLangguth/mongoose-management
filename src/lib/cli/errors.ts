/* tslint:disable:max-classes-per-file */

export class CancelPromptError extends Error {}

export class BackToCollectionError extends Error {}
