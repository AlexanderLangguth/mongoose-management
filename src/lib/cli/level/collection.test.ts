jest.mock('../../prompts');
jest.mock('./column');
jest.mock('./index');

import Prompts from '../../prompts';
import CollectionDataset from '../dataset/collection';
import ColumnDataset from '../dataset/column';
import GroupDataset from '../dataset/group';
import IndexDataset from '../dataset/index';
import { BackToCollectionError } from '../errors';
import CollectionMenu from '../menu/collection';
import promptsCollection from '../prompts/collection';
import ColumnLevel from './column';
import IndexLevel from './index';

import CollectionLevel from './collection';

import { dataColumnType } from '../../types';

describe('Check the CollectionLevel class', () => {
  let prompts: any;
  let storage: any;
  let creater: any;
  let group: GroupDataset;
  let dataset: CollectionDataset;
  let level: any;

  beforeEach(() => {
    prompts = new Prompts();
    storage = jest.fn();
    creater = jest.fn();

    group = new GroupDataset({ path: 'path/to/project', collections: [] }, jest.fn() as any);
    dataset = new CollectionDataset(
      {
        name: 'collectionName',
        columns: [
          { name: 'c1', type: 'string' },
          { name: 'c2', type: 'object' },
        ],
        indexes: [{ name: 'i1', columns: { c1: 'hashed' }, properties: {} }],
      },
      group,
    );
    group.addCollection(dataset);

    level = new CollectionLevel(dataset, { prompts, storage, creater });
  });

  test('initialize the class', () => {
    expect(level).toBeInstanceOf(CollectionLevel);

    expect(level.dataset).toBeInstanceOf(CollectionDataset);
    expect(level.menu).toBeInstanceOf(CollectionMenu);
    expect(level.prompts).toBe(prompts);
    expect(level.options).toEqual({ prompts, storage, creater });

    expect(level.promptCreate).toEqual(expect.any(Function));
    expect(level.promptEdit).toBe(promptsCollection);
  });

  test('it should be promptsIndex() called when create() is called with action "createIndex"', async () => {
    const c1 = dataset.getColumn('c1');

    ((prompts.call as any) as jest.Mock).mockResolvedValueOnce({ name: 'indexName', columns: [c1] });
    ((prompts.call as any) as jest.Mock).mockResolvedValueOnce({ c1: [c1, 1] });
    ((prompts.call as any) as jest.Mock).mockResolvedValueOnce({ unique: true, sparse: false });

    expect(dataset.getIndexes().length).toBe(1);

    const result = await level.create('createIndex');

    expect(result).toBe(undefined);

    expect(dataset.getIndexes().length).toBe(2);
    expect(prompts.call).toHaveBeenCalledTimes(3);
    expect(prompts.call).toHaveBeenCalledWith(expect.any(Array));
  });

  test('it should be promptsIndex() called when create() is called with action "createColumn" and type "string"', async () => {
    ((prompts.call as any) as jest.Mock).mockResolvedValueOnce({ name: 'columnName', type: 'string' });
    ((prompts.call as any) as jest.Mock).mockResolvedValueOnce({ options: ['default'], default: "'Moin'" });
    ((prompts.call as any) as jest.Mock).mockResolvedValueOnce({ type: 'unique', value: 'hashed' });

    expect(dataset.getColumns().length).toBe(5);

    const result = await level.create('createColumn');

    expect(result).toEqual(undefined);

    expect(dataset.getColumns().length).toBe(6);
    expect(prompts.call).toHaveBeenCalledTimes(3);
    expect(prompts.call).toHaveBeenCalledWith(expect.any(Array));
  });

  test.each([['array'], ['object']])(
    'it should be promptsIndex() called when create() is called with action "createColumn" and type "%s"',
    async (type) => {
      ((prompts.call as any) as jest.Mock).mockResolvedValueOnce({ name: 'columnName', type });

      expect(dataset.getColumns().length).toBe(5);

      const result = await level.create('createColumn');

      expect(result).toEqual(expect.any(ColumnDataset));

      expect(dataset.getColumns().length).toBe(6);
      expect(prompts.call).toHaveBeenCalledTimes(1);
      expect(prompts.call).toHaveBeenCalledWith(expect.any(Array));
    },
  );

  test('it should be throw an error when create() is called with unknwon action', async () => {
    expect.assertions(2);
    try {
      await level.create('create');
    } catch (err) {
      expect(err).toBeInstanceOf(Error);
      expect(err.message).toBe('Unknown action');
    }
  });

  test('it should be remove the collection when remove() is called', async () => {
    (prompts.remove as jest.Mock).mockImplementation(async (name: any) => {
      expect(typeof name).toBe('string');

      return true;
    });

    expect(group.getCollections().length).toBe(1);

    const result = await level.remove(dataset);

    expect(result).toBe(false);

    expect(group.getCollections().length).toBe(0);
    expect(prompts.remove).toHaveBeenCalledTimes(1);
  });

  test('it should be throw an error when remove() is called with a column reference', async () => {
    const columns: dataColumnType[] = [{ name: 'column2', type: 'objectId', populate: 'collectionName' }];
    group.addCollection(new CollectionDataset({ name: 'collection2', columns, indexes: [] }, group));

    expect.assertions(5);
    expect(group.getCollections().length).toBe(2);

    try {
      await level.remove(dataset);
    } catch (err) {
      expect(err).toBeInstanceOf(Error);
      expect(err.message).toMatchSnapshot();

      expect(group.getCollections().length).toBe(2);
      expect(prompts.remove).toHaveBeenCalledTimes(0);
    }
  });

  test('it should be create IndexLevel when show() is called with an IndexDataset', async () => {
    const subDataset = dataset.getIndex('i1');

    expect(subDataset).toBeInstanceOf(IndexDataset);

    await level.show(subDataset);

    expect(IndexLevel).toHaveBeenCalledTimes(1);
    expect(IndexLevel).toHaveBeenCalledWith(subDataset, { prompts, storage, creater });
    expect(IndexLevel.prototype.exec).toHaveBeenCalledTimes(1);
    expect(IndexLevel.prototype.exec).toHaveBeenCalledWith();
  });

  test('it should be create ColumnLevel when show() is called with a ColumnDataset', async () => {
    const subDataset = dataset.getColumn('c1');

    expect(subDataset).toBeInstanceOf(ColumnDataset);

    await level.show(subDataset);

    expect(ColumnLevel).toHaveBeenCalledTimes(1);
    expect(ColumnLevel).toHaveBeenCalledWith(subDataset, { prompts, storage, creater });
    expect(ColumnLevel.prototype.exec).toHaveBeenCalledTimes(1);
    expect(ColumnLevel.prototype.exec).toHaveBeenCalledWith();
  });

  test('it should be catch the error BackToCollectionError when show() is called and this one throws the error BackToCollectionError', async () => {
    (ColumnLevel.prototype.exec as jest.Mock).mockRejectedValue(new BackToCollectionError('back to collection'));

    const subDataset = dataset.getColumn('c1');

    expect(subDataset).toBeInstanceOf(ColumnDataset);

    await level.show(subDataset);

    expect(ColumnLevel).toHaveBeenCalledTimes(1);
    expect(ColumnLevel).toHaveBeenCalledWith(subDataset, { prompts, storage, creater });
    expect(ColumnLevel.prototype.exec).toHaveBeenCalledTimes(1);
    expect(ColumnLevel.prototype.exec).toHaveBeenCalledWith();
  });

  test('it should be forwards the error when show() is called and this one throws an error', async () => {
    expect.assertions(7);

    (ColumnLevel.prototype.exec as jest.Mock).mockRejectedValue(new Error('other error'));

    const subDataset = dataset.getColumn('c1');

    expect(subDataset).toBeInstanceOf(ColumnDataset);

    try {
      await level.show(subDataset);
    } catch (err) {
      expect(err).toBeInstanceOf(Error);
      expect(err.message).toBe('other error');
    }

    expect(ColumnLevel).toHaveBeenCalledTimes(1);
    expect(ColumnLevel).toHaveBeenCalledWith(subDataset, { prompts, storage, creater });
    expect(ColumnLevel.prototype.exec).toHaveBeenCalledTimes(1);
    expect(ColumnLevel.prototype.exec).toHaveBeenCalledWith();
  });

  test('it should be throw an error when show() is called without Dataset', async () => {
    expect.assertions(6);
    try {
      await level.show({});
    } catch (err) {
      expect(err).toBeInstanceOf(Error);
      expect(err.message).toBe('Unknown dataset instance');
    }

    expect(IndexLevel).toHaveBeenCalledTimes(0);
    expect(IndexLevel.prototype.exec).toHaveBeenCalledTimes(0);
    expect(ColumnLevel).toHaveBeenCalledTimes(0);
    expect(ColumnLevel.prototype.exec).toHaveBeenCalledTimes(0);
  });

  test('it should be throw an error when promptCreate() is called', async () => {
    expect.assertions(2);
    try {
      await level.promptCreate();
    } catch (err) {
      expect(err).toBeInstanceOf(Error);
      expect(err.message).toBe('Creating action is not done in the abstract class');
    }
  });
});
