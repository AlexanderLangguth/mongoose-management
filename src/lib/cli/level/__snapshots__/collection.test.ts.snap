// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the CollectionLevel class it should be throw an error when remove() is called with a column reference 1`] = `
"This collection is referenced by other column(s):

- [1mcollection2[22m.column2

The references must first be deleted"
`;
